chore: update rekordbox kaitai definitions

Swiftb0y opened the pull request:
Should also fix #13291.

Lets see if this builds without updating any additional changes.

Swiftb0y committed:
chore: update rekordbox kaitai definitions

Swiftb0y commented:
Apparently not, seems like we need an updated kaitaistruct runtime. It also doesn't seem appropriate to vendor that anymore. Would `ExternalProject_Add` instead be an option? Can somebody help me with the required CMake legwork for that?

JoergAtGithub commented:
We have it in VCPKG

Swiftb0y commented:
yeah, but thats about it, its not packaged on any linux distro. https://repology.org/project/kaitai-struct-cpp-stl-runtime/versions

daschuer commented:
We cannot download external sources on Launchpad. So let's just update our vendored version.

Swiftb0y committed:
update kaitai cpp stl runtime to v0.10.1

Swiftb0y commented on lib/kaitai/kaitai/kaitaistream.h:
```diff
@@ -162,7 +167,11 @@ class kstream {
     static std::string bytes_terminate(std::string src, char term, bool include);
-    static std::string bytes_to_str(std::string src, std::string src_enc);
+    static std::string bytes_to_str(const std::string src, const char *src_enc);
+    // workaround for https://github.com/kaitai-io/kaitai_struct_cpp_stl_runtime/issues/67
+    static std::string bytes_to_str(const std::string src, const std::string& enc) {
+      return bytes_to_str(src, enc.c_str());                       
+    }
```
only changed I made to the kaitai files except copy-pasting in the newest commit and moving them to match the includes the kaitai-compiler generates.

Swiftb0y commented on src/library/rekordbox/kaitaistructs/rekordbox_anlz.ksy:
```diff
@@ -393,77 +411,151 @@ types:
+          The rest of the tag, which needs to be unmasked before it
+          can be parsed.
+        size-eos: true
+        # NOTE: unmasking is disabled because the C++ backend of kaitai_struct
+        # doesn't seem to support typecasting as of kaitai version 0.10?
+        # process:  'xor(is_masked ? mask : [0])'
+    instances:
+      c:
+        value: len_entries
+      # mask:
+      #   value: |
+      #     [
+      #       (0xCB+c).as<u1>, (0xE1+c).as<u1>, (0xEE+c).as<u1>, (0xFA+c).as<s1>, (0xE5+c).as<s1>, (0xEE+c).as<s1>, (0xAD+c).as<s1>, (0xEE+c).as<s1>,
+      #       (0xE9+c).as<u1>, (0xD2+c).as<u1>, (0xE9+c).as<u1>, (0xEB+c).as<s1>, (0xE1+c).as<s1>, (0xE9+c).as<s1>, (0xF3+c).as<s1>, (0xE8+c).as<s1>,
+      #       (0xE9+c).as<u1>, (0xF4+c).as<u1>, (0xE1+c).as<u1>
+      #     ].as<bytes>
+      raw_mood:
```
see comment

Swiftb0y commented on src/library/rekordbox/rekordboxfeature.cpp:
```diff
@@ -523,82 +524,57 @@ QString parseDeviceDB(mixxx::DbConnectionPoolPtr dbConnectionPool, TreeItem* dev
-                ++table) {
-            if ((*table)->type() == tableOrder[tableOrderIndex]) {
-                uint16_t lastIndex = (*table)->last_page()->index();
-                rekordbox_pdb_t::page_ref_t* currentRef = (*table)->first_page();
+        for (const auto& table : *rekordboxDB.tables()) {
+            if (table->type() == tableOrder[tableOrderIndex]) {
+                uint16_t lastIndex = table->last_page()->index();
+                rekordbox_pdb_t::page_ref_t* currentRef = table->first_page();
```
this didn't build anymore because the `table_t` changed to a unique_ptr. I took the liberty and modernized the loop in the same step (same below).

Swiftb0y commented:
@0xsuk can you [test this PR](https://github.com/mixxxdj/mixxx/wiki/Testing#github-pull-requests) to see whether this doesn't break anything and fixes your string-encoding issue? Thank you.

Swiftb0y committed:
make kaitai rekordbox code a library

This makes it easier to ignore warnings in the autogenerated code.
It also makes more sense since the `*.ksy` files are effectively
vendored and the generated C++ files are autogenerated. Both
something that doesn't belong in `src/`.

0xsuk commented:
Everything looks good after I changed UTF-16BE to UTF-16LE in src/library/rekordbox/rekordboxfeature.cpp line 259.

0xsuk commented:
![image](https://github.com/mixxxdj/mixxx/assets/97814789/f4f6c6ef-e4b1-4a77-b4b4-150861977389)
Thank you :)

Swiftb0y commented:
whoops. good catch, I assumed kaitai would take care of the conversion automatically. Fixed

Swiftb0y commented:
Ready for review. I suggest to go through on a commit by commit basis. The PR looks bigger as it is because most of the changes are either auto-generated or artifacts because I moved files around.

Swiftb0y commented:
friendly ping, would be nice if this made it into 2.4.2. I've highlighted the significant changes above.

Swiftb0y commented on CMakeLists.txt:
```diff
@@ -2508,12 +2486,20 @@ target_link_libraries(mixxx-lib PRIVATE FpClassify)
 target_link_libraries(mixxx-lib PRIVATE mp3lame::mp3lame)
 
+add_library(rekordbox_metadata STATIC EXCLUDE_FROM_ALL
+  lib/rekordbox-metadata/rekordbox_pdb.cpp
+  lib/rekordbox-metadata/rekordbox_anlz.cpp
+)
+target_include_directories(rekordbox_metadata SYSTEM PUBLIC lib/rekordbox-metadata)
+target_link_libraries(mixxx-lib PRIVATE rekordbox_metadata)
+
 # Kaitai for reading Rekordbox libraries
 add_library(Kaitai STATIC EXCLUDE_FROM_ALL
-  lib/kaitai/kaitaistream.cpp
+  lib/kaitai/kaitai/kaitaistream.cpp
 )
```
I got a lot of code quality complaints on the output of the kaitai compiler, so I moved the generated code to `lib/` (first and foremost because the linters already ignore that and second because it also makes sense to treat autogenerated code that way).

Swiftb0y committed:
rekordbox: incorporate https://github.com/Deep-Symmetry/crate-digger/issues/32 fix

daschuer reviewed (approved):
The code looks good and I rely on the tests done by @0xsuk
Thank you.

daschuer commented:
How important is the fix? Does this rectify a release?

Swiftb0y commented:
Thank you. 

> How important is the fix? Does this rectify a release?

Well this PR contains two fixes, the string decoding issue reported in #13291. The second one concerns history playlists created by rekordbox. I don't even think we ingest those.

I don't think an extra release is warranted as only the string decoding issue is affected, judging by the delay with which this was reported, I wouldn't say its critical enough to justify the effort.

brunchboy commented:
It’s not just history playlists. Any table which has had deletions over time will likely be missing a few rows at the end without the structure fix.

brunchboy commented:
And not just deletions, rows are sometimes not present in row groups for inscrutable reasons. The fact that the problem was noticed in the history playlist table shows it could happen anywhere. We need to scan the entire last row group for tracks marked as present, not assume we can stop scanning based on the value of `num_rows`.

brunchboy commented:
Which is not to say that it necessarily merits an urgent release, I have no idea how many people rely on scanning rekordbox USBs in the first place. 😄

Swiftb0y commented:
Ah, thank you. In that case it would probably justify a release. Considering there is also #13309 #11373 and #10923 which all could be caused by that issue.

brunchboy commented:
Hmm, those all could be impacted but they sound worse than this! Please feel free to ping me if this doesn’t fix those other issues and perhaps I can help investigate the problematic export files.

Swiftb0y commented:
That would be wonderful. We can close these issues once we have a release with the fix and ask the respective issue authors to check again. In case it still happens, I would refer them to you (where exactly? crate-digger repo / deep-symmetry zulip) if thats alright for you.

brunchboy commented:
Let’s start on the Zulip instance and then I can create issues as appropriate.
